### ISSUE TITLE
Nested contents?

### ISSUE THREAD

**Issue by binarykitchen** [keystonejs/keystone#646]
It would be great if Keystone allows nested contents, stacked as if these were modules. For a blog entry my client wants to assemble contents by pieces from other contents (1 .. n relationship).

Let me rephrase the request:
Each of these modules would have some text and images, each of them might be aligned differently. How can I create contents based on those modules?

The reason for this request is because these modules have to re-appear a couple of times among the whole page, within different content holders.

**Comment by morenoh149** [comment]
@binarykitchen could you submit a pull request?

**Comment by BerkeleyTrue** [comment]
@binarykitchen Is this a duplicate of https://github.com/keystonejs/keystone/issues/257 ? If so, please close.

**Comment by JedWatson** [comment]
I actually suspect Relationships (or some variation) may be the right feature for this.

You could set up `Modules` as a list, then let the editors include them via relationships in the appropriate content holders (i.e. you have your content holders set up as `Relationship` fields on the page model...

The Admin UI might need a bit of work to make this a good experience for users (might be able to use `dependsOn` and `filters` to make it more usable) - I'm guessing a bit here as to what you want to do.

@binarykitchen maybe if you could mock up a screenshot of what you've got in mind it would help us understand it more clearly.

**Comment by iamkevingreen** [comment]
@JedWatson I've actually imagined this for a long time with Keystone. There are a few tools that work with meteor that do this type of schema building and it's similar to how you have Relationship Arrays set up only the items within the array are repeatable. 

I think instead of a mockup something like this is an ideal example: http://www.advancedcustomfields.com/add-ons/flexible-content-field/

Basically if we could define blocks that would be repeatable and placed into an array that we could iterate over on the frontend. 

So maybe it's a relationship field and we can define blocks that have things like, 2 column layout that has 2 wyiswyg editors, or a full width block that just has an image upload and you can stack them however you like. 

I've also been working with Meteor and the Aldeed schema package handles it really nicely which you can demo here(also on mongo): http://autoform.meteor.com/quickform if you click the 'array of object fields' for example

You could even have all of the fields live inside the content Object and have a dropdown that just has a bunch of options that dependsOn talks to. But again I think the idea of repeatable items is what is lacking in the current schema builder. 

I believe this is what @binarykitchen had in mind.

**Comment by msimpson** [comment]
What @iamkevingreen implied that @binarykitchen had in mind is exactly what I was looking for as well. Without some sort of sub-collection or nested content support, even the most simplistic of dynamic content challenges results in an absurdly difficult solution for end users. Such as a multiple model and relationship concoction to tackle a dynamic carousel with text captions.

**Comment by medv** [comment]
Silverstripe CMS solves this pretty well using their GridField component. http://www.silverstripe.org/software/download/

**Comment by creynders** [comment]
We're closing all feature requests to keep the issue tracker unpolluted. I've moved this to ProductPains:
https://productpains.com/post/keystonejs/add-support-for-nested-repeating-schemas

**Comment by msimpson** [comment]
586 days and the only resolution is to move these requests to an essentially unknown site so they can become lost in an unorganized list of "product pains." Right ...

**Comment by mxstbr** [comment]
@msimpson sorry for the troubles, but this is a very positive step for us as maintainers. We had 589 open issues, which made it impossible to have a clear picture of priorities. We decided to to focus this ISSUE tracker on bugs, and use a site made for the purpose of collection feature requests for said requests.

This "essentially unknown site" is maintained by a good friend of mine, whom you can [easily find and contact](https://twitter.com/a13n). I spoke to him before doing this, and made sure that longevity was part of his plan. ProductPains isn't going anywhere anytime soon.

Calling ProductPains "an unorganized list" when _comparing it against the Github issue tracker_ is quite the statement. Contrary to what we had before we now have a **clear picture of which features our users [want most](https://productpains.com/product/keystonejs?tab=top)**.
One example of something we weren't aware of was how many users want LetsEncrypt support, so that's something we'll focus on as soon as possible. We also couldn't have been aware, because there only was a single open issue about that.

Again, we're sorry for the troubles this will cause in this (hopefully short) moving period, but in the long run this is an important step towards making Keystone as stable and good as we can.

**Comment by creynders** [comment]
@msimpson this is an open source project, if the development of it (or a specific feature) doesn't happen quite speedily enough in your opinion you're more than welcome to help out a hand. And, sneering people creating a product for you to use for free is in pretty bad taste. What did _you_ contribute to Keystone in those 586 days?

**Comment by andrewlinfoot** [comment]
@msimpson thank you for your input regarding our migration to product pains. we are sorry that you do not like this solution and are concerned that this issue may become "lost in an unorganized list". Seeing as this is such a deep concern for you, we would love if you could implement your ideal solution and submit a pull request. We will make sure to review it and either merge or provide feedback in a timely manner.

**Comment by msimpson** [comment]
@mxstbr @creynders @andrewlinfoot 
"What did you contribute to Keystone in those 586 days?"
That's easy. Nothing! Because Keystone doesn't even support simple features like nested contents.

**Comment by mxstbr** [comment]
No further comments, this is obviously a troll.
